fix error with string initialisation in plot.c  & linker updates

mattiamtzlr opened the pull request:
In the functions `plot_function` and `plot_implicit`, the string initialisation `static char c[1] = "o"` causes an error as the array of size 1 does not provide space for the NUL-terminator. Thus the fix is to increase the array size to 2: `static char c[2] = "o"`.

In addition, I updated the header files to use the `#pragma once` directive for easier include management.

Finally, I added the linker flag `-z muldefs` to allow multiple definitions with arise from the globally defined variables in the header files. This should maybe be redesigned at some point. Also, I updated the C standard to c23.

Cool project! :)

mattiamtzlr committed:
update ifndef checks to use pragma once

mattiamtzlr committed:
fix string initialisation in plot.c

mattiamtzlr committed:
allow multiple definitions in linker, update C standard

gciruelos commented:
Hi Mattia,

>In the functions plot_function and plot_implicit, the string initialisation static char c[1] = "o" causes an error as the array of size 1 does not provide space for the NUL-terminator. Thus the fix is to increase the array size to 2: static char c[2] = "o".

LGTM

>In addition, I updated the header files to use the #pragma once directive for easier include management.

LGTM

>Finally, I added the linker flag -z muldefs to allow multiple definitions with arise from the globally defined variables in the header files. This should maybe be redesigned at some point. Also, I updated the C standard to c23.

* Can you explain what `-z multdefs` does exactly for the project today? IIUC this is meant to avoid errors that arise from having the same symbol defined multiple times. This should not be the case today as we link either termbox or ncurses, but not both.

* Also, unfortunately I'm running an old version of gcc (9.4). Can you either change the std to `c17` or `c2x`? These have been supported for a while. 

I will merge once these questions / issues are resolved. Thank you for your contribution!

gciruelos commented:
>Also, unfortunately I'm running an old version of gcc (9.4). Can you either change the std to c17 or c2x? These have been supported for a while.

Actually, I don't think this is a reasonable ask from my side. Forget about this comment (I'm updating my distro), and just address the -z multdefs one.

mattiamtzlr commented:
Hi Gonzalo, thanks a lot for your reply!  

> Can you explain what` -z multdefs` does exactly for the project today? IIUC this is meant to avoid errors that arise from having the same symbol defined multiple times. This should not be the case today as we link either termbox or ncurses, but not both.

When I try to build the project without this linker flag, I get the following output:
```
cc -Wall -Wextra -Werror -Wshadow -Wstrict-prototypes -Wpointer-arith -Wcast-qual -Werror -std=c17 -pedantic -DINCL_TERMBOX -O3 -flto -ffast-math -c -o obj/command.o src/command.c
cc -O3 -flto -ffast-math obj/command.o obj/debug.o obj/options.o obj/parser.o obj/plot.o obj/termplot.o obj/ui.o obj/ui_termbox.o -o termplot -lm  -ltermbox
/usr/lib64/gcc/x86_64-suse-linux/15/../../../../x86_64-suse-linux/bin/ld: obj/debug.o (symbol from plugin): in function `start_debug':
(.text+0x0): multiple definition of `debug_enabled'; obj/command.o (symbol from plugin):(.text+0x0): first defined here
/usr/lib64/gcc/x86_64-suse-linux/15/../../../../x86_64-suse-linux/bin/ld: obj/options.o (symbol from plugin): in function `init_structs':
(.text+0x0): multiple definition of `options'; obj/command.o (symbol from plugin):(.text+0x0): first defined here
/usr/lib64/gcc/x86_64-suse-linux/15/../../../../x86_64-suse-linux/bin/ld: obj/parser.o (symbol from plugin): in function `_neg':
(.text+0x0): multiple definition of `debug_enabled'; obj/command.o (symbol from plugin):(.text+0x0): first defined here
[...]
/usr/lib64/gcc/x86_64-suse-linux/15/../../../../x86_64-suse-linux/bin/ld: obj/ui.o (symbol from plugin): in function `update_cmd':
(.text+0x0): multiple definition of `functions'; obj/command.o (symbol from plugin):(.text+0x0): first defined here
/usr/lib64/gcc/x86_64-suse-linux/15/../../../../x86_64-suse-linux/bin/ld: obj/ui.o (symbol from plugin): in function `update_cmd':
(.text+0x0): multiple definition of `debug_enabled'; obj/command.o (symbol from plugin):(.text+0x0): first defined here
collect2: error: ld returned 1 exit status
make: *** [Makefile:61: termplot] Error 1
```
with more of the same at the [...], which AFAIU suggests that many symbols of the `command.o` object file have multiple definitions. My LSP tells me that the variables defined in the header files (such as `command` in `command.h`) can lead to one-definition-rule violations, so I think that's probably the reason for these errors.
I agree that this is weird and seems unneeded, as AFAIK there aren't any actual re-definitions, but I could not get it to build otherwise...

Hope this somewhat clears things up :)

gciruelos commented:
Thank you Mattia.

The proper fix is a bit longer (but easy), I'd suggest you attempt it. I tried it locally and it works. Reference: https://en.cppreference.com/w/c/language/extern.html

Essentially you need, for each global variable in an .h file, to declare it as `extern` and set it in the corresponding `.c` file.

```
// debug.h
extern int debug_enabled;
``` 

```
// debug.c
int debug_enabled = 0;
```

```
// options.h
extern struct options_t options;
```

```
// options.c
struct options_t options;
``` 


```
// plot.h
struct function {
    expr f;
    expr g;
    unsigned char type;
    unsigned char valid;
};

extern struct function functions[MAX_FUNCTIONS];
```

```
// plot.c
struct function functions[MAX_FUNCTIONS];
```

```
// command.h

typedef struct commant_history_t {
  // Kind of bounded stack.
  char* commands[CMD_HIST];
  unsigned int last;
  unsigned int first;
} command_history_t;

extern command_history_t command_history;

extern char command[CMD_SIZE];
extern unsigned int cursor;
```

```
// command.c
command_history_t command_history;
char command[CMD_SIZE];
unsigned int cursor = 0;
```

```
// ui_curses.h
extern WINDOW* win;
extern bool curses_started;
```

```
// ui_curses.c

bool curses_started = false;
WINDOW* win = NULL;
```


This builds on my PC. I cannot accept `-z muldefs` because it does not fix the problem, it just masks it, which can cause problems down the line. Let me know if you have problems adapting the patch and I can help.

Thanks again for your contribution.

mattiamtzlr committed:
actually fix variable definitions in header file instead of hacking it

uses extern declarations and subsequent definitions, instead of passing
the muldefs flag to the linker

mattiamtzlr commented:
Hi again,
You're completely right, I thought about using `extern` but deemed it too complicated, but it now just took me 10 minutes to implement it :) Now everything seems to be working well!
Thanks for your feedback!
